src/Bus/MongoBatchRepository.php

```diff
@@ -0,0 +1,270 @@
+        $this->collection->updateOne(
+            ['_id' => $batchId],
+            [
+                '$set' => [
+                    'cancelled_at' => new UTCDateTime(Carbon::now()),
+                    'finished_at' => new UTCDateTime(Carbon::now()),
+                ],
+            ],
+        );
+    }
+
+    #[Override]
+    public function delete(string $batchId): void
+    {
+        $batchId = new ObjectId($batchId);
+        $this->collection->deleteOne(['_id' => $batchId]);
+    }
+
+    /** Execute the given Closure within a storage specific transaction. */
+    #[Override]
+    public function transaction(Closure $callback): mixed
+    {
+        return $this->connection->transaction(fn () => $callback());
+    }
+
+    /**
+     * Rollback the last database transaction for the connection.
+     */
+    #[Override]
+    public function rollBack(): void
+    {
+        $this->connection->rollBack();
+    }
+
+    /** Mark the batch that has the given ID as finished. */
+    #[Override]
+    public function prune(DateTimeInterface $before): int
+    {
+        $result = $this->collection->deleteMany(
+            ['finished_at' => ['$ne' => null, '$lt' => new UTCDateTime($before)]],
+        );
+
+        return $result->getDeletedCount();
+    }
+
+    /** Prune all the unfinished entries older than the given date. */
+    public function pruneUnfinished(DateTimeInterface $before): int
+    {
+        $result = $this->collection->deleteMany(
+            [
+                'finished_at' => null,
+                'created_at' => ['$lt' => new UTCDateTime($before)],
+            ],
+        );
+
+        return $result->getDeletedCount();
+    }
+
+    /** Prune all the cancelled entries older than the given date. */
+    public function pruneCancelled(DateTimeInterface $before): int
+    {
+        $result = $this->collection->deleteMany(
+            [
+                'cancelled_at' => ['$ne' => null],
+                'created_at' => ['$lt' => new UTCDateTime($before)],
+            ],
+        );
+
+        return $result->getDeletedCount();
+    }
+
+    #[Override]
+    protected function toBatch($batch): Batch
+    {
+        return $this->factory->make(
+            $this,
+            $batch['_id'],
+            $batch['name'],
+            $batch['total_jobs'],
+            $batch['pending_jobs'],
+            $batch['failed_jobs'],
+            $batch['failed_job_ids'],
+            unserialize($batch['options']),
+            $this->toCarbon($batch['created_at']),
+            $this->toCarbon($batch['cancelled_at']),
+            $this->toCarbon($batch['finished_at']),
+        );
+    }
+
+    /** @return ($date is null ? null : CarbonImmutable) */
+    private function toCarbon(?UTCDateTime $date): ?CarbonImmutable
+    {
+        if ($date === null) {
+            return null;
+        }
+
+        return CarbonImmutable::createFromTimestamp((string) $date, date_default_timezone_get());
```

Comment:
AFAIK, `createFromTimestamp()` and `createFromTimestampUTC()` both expect the argument to denote the timestamp in seconds. [`UTCDateTime::__toString()`](https://www.php.net/manual/en/mongodb-bson-utcdatetime.tostring.php) returns a numeric string denoting milliseconds.

Would `createFromTimestampMsUTC()` be the correct method to use here? I'm also concerned about _not_ using the "UTC" variant for the Carbon method here, as that might blindly apply the current system timezone to the timestamp. The Carbon docs aren't clear about _how_ the optional timezone gets applied to the timestamp (ideally, it would always assume a timestamp is UTC and apply a conversion).